### ISSUE TITLE
Derive Table of Contents for Guides/Articles

### ISSUE THREAD

**Issue by martinklepsch** [cljdoc/cljdoc#20]
Individual pages may be long, having multiple sections etc. It would be nice if there were some useful UI elements that would help with navigating long pages. This could be part of the sidebar, a Table of Contents at the top of documents or something else entirely. 

The following is from an early mockup, note the "Current Section" element:

<img width="801" alt="cljdoc" src="https://user-images.githubusercontent.com/97496/38617289-e95d8ad4-3d96-11e8-90b0-7e2aa1c8d16e.png">

**Comment by eerohele** [comment]
For Markdown, guess one option would be to write a [CommonMark Visitor] to extract the headings from the Markdown files and include them in the doc tree.

**Comment by martinklepsch** [comment]
Yeah that would be an option. We’d then need something different for Asciidoc though which is why I am thinking that processing the HTML output of those might be the more flexible approach. 

I played around with Jsoup for that kind of thing (`cljdoc.util.fixref`) and it seems to be working fairly well.

**Comment by martinklepsch** [comment]
After a bit more thought I think the following would be a worthwhile approach:

- Write some JS that extracts headings and their IDs from a provided `<div>`
- Write some more JS that takes a structure like this and renders an interactive component of sorts with all headings in them and clickable
  - That component should be collapsible (see "Current section" bit in screenshot above)
  - We might want to restrict to what level of depth headings are shown
  - We might want to ignore the highest-level heading if it occurs only once (e.g. if there's one `<h1>` there's usually no point in showing it.)
- Write more JS to monitor scroll position and update the "Curren section"/highlighted section 

I'll try to create some mockups for this later

**Comment by martinklepsch** [comment]
@arrdem I assume you meant stacks? :D The issue with anything tied to CommonMark is that it also ties us to Markdown.

For this particular situation the information is very compact, i.e. we really just need the document contents and nothing else to do the processing. We could do this in either CLJ or JS but I feel like if we want to get the UX right we'll end up with some JS component either way so why not just skip any server side stuff...

**Comment by martinklepsch** [comment]
Here are some more thoughts on how I imagine this "collapsible table of contents" to work:

---
**Collapsed state:**
![toc collapsed](https://user-images.githubusercontent.com/97496/45926056-3c3c1380-bf27-11e8-8232-a73b5b67ffd3.png)

---
**Expanded state:**
![toc expanded](https://user-images.githubusercontent.com/97496/45926057-3e9e6d80-bf27-11e8-9418-560ee486ff18.png)

--- 
**More notes:**

- I've been using Preact for some JS parts like the search on the homepage and the (still somewhat hidden) quick switcher ([code](https://github.com/cljdoc/cljdoc/blob/master/resources/public/switcher.js)). Since this will also have some state it might make sense to encapsulate it in a component too.
- I think we should left align the TOC bar since long titles will make it look off-balance.
- As for what levels of headings to include: Without much testing I think only displaying the highest level that has more than one occurrence could be good (e.g. `h2` in [Data-driven middleware](https://cljdoc.xyz/d/metosin/reitit/0.1.3/doc/ring/data-driven-middleware))
- There's still a few unknowns about the expanded state (added them as questions to the image) where I'd suggest we just do what feels easiest for now.

**Comment by zcaudate** [comment]
elixir has really nice docs

https://hexdocs.pm/elixir/master/Kernel.html

**Comment by martinklepsch** [comment]
Hexdocs is nice indeed. Although I have to say I find the hierarchy in the sidebar fairly confusing  (Disclaimer: I don't know a think about Elixir 😄)

**Comment by zcaudate** [comment]
@martinklepsch: yeah, I'm not a big fan that either (the hierarchy thing for api).  

The thing with elixir is that they've pretty much made docs first class within the language and so can mix apis and docs like that. It does makes things a little bit confusing.

The collapsible TOC is great. One suggestion:
- for mobile devices, it is at the top
- for desktop/larger screens, it floats on the right.

**Comment by lread** [comment]
A nice example from @NoahTheDuke on Slack. He'd use this new feature for this page:
https://cljdoc.org/d/io.github.noahtheduke/splint/1.20.0/doc/rules-overview/style

![Image](https://github.com/user-attachments/assets/b4b86045-5495-4fc3-93e5-9c74a3afd66d)

**Comment by lread** [comment]
Hmm... I did not entirely notice/remember that GitHub now has a toc feature:

![Image](https://github.com/user-attachments/assets/46fdede6-b898-4b2b-991e-3d51ca960474)

![Image](https://github.com/user-attachments/assets/b475cae4-9cc8-489f-b406-91a8ccc1202d)

I don't know if we'll get inspiration from this design, but I was partly generating tocs in my articles for readers who use GitHub. So, interesting!

**Comment by Ramblurr** [comment]
@lread I'd love to help push this idea forward, including with code. Looking at the history since 2018, it seems we need a concrete feature list and then a visual design proposal?

1. Which heading levels should be included? Should that be configurable?
2. How should the TOC be laid out on desktop and mobile?
3. What should library authors be able to configure?
4. Should article TOC and vars nav share behavior/styling?
5. Which formats? (markdown and adoc.. txt?)

Some of these have simple answers, others might need a visual mockup.

If you have any strong opinions (on these or other questions), I'd like to hear them. Otherwise, I'd start working on a little mock-up.

**Comment by lread** [comment]
That's great @Ramblurr! Looks like you've got some great questions there!

I think we'd support all formats. But .txt doesn't have headings. So, nothing we can do for .txt, I think.

And I think you are proposing doing this at the doc level? There might be some discussion above doing it at the article TOC level.

**Comment by Ramblurr** [comment]
> There might be some discussion above doing it at the article TOC level.
> 

I'm not sure what you mean with that distinction? Could you clarify?

**Comment by lread** [comment]
Yeah, good stuff. Do you have a preference?
